ddtrace/contrib/utils.py

```diff
@@ -0,0 +1,143 @@
+    """TracedCursor for psycopg instances"""
+
+    def __init__(self, cursor, pin, cfg, *args, **kwargs):
+        super(PsycopgTracedCursor, self).__init__(cursor, pin, cfg)
+
+    def _trace_method(self, method, name, resource, extra_tags, dbm_propagator, *args, **kwargs):
+        # treat Composable resource objects as strings
+        if resource.__class__.__name__ == "SQL" or resource.__class__.__name__ == "Composed":
+            resource = resource.as_string(self.__wrapped__)
+        return super(PsycopgTracedCursor, self)._trace_method(
+            method, name, resource, extra_tags, dbm_propagator, *args, **kwargs
+        )
+
+
+class PsycopgFetchTracedCursor(PsycopgTracedCursor, dbapi.FetchTracedCursor):
+    """PsycopgFetchTracedCursor for psycopg"""
+
+
+class PsycopgTracedConnection(dbapi.TracedConnection):
+    """PsycopgTracedConnection wraps a Connection with tracing code."""
+
+    def __init__(self, conn, pin=None, cursor_cls=None):
+        package = conn.__class__.__module__.split(".")[0]
+        if not cursor_cls:
+            # Do not trace `fetch*` methods by default
+            cursor_cls = (
+                PsycopgFetchTracedCursor if config._config[package].trace_fetch_methods else PsycopgTracedCursor
+            )
+
+        super(PsycopgTracedConnection, self).__init__(conn, pin, config._config[package], cursor_cls=cursor_cls)
+
+
+def patch_conn(conn, traced_conn_cls=PsycopgTracedConnection, pin=None):
+    """Wrap will patch the instance so that its queries are traced."""
+    # ensure we've patched extensions (this is idempotent) in
+    # case we're only tracing some connections.
+    _config = None
+    if pin:
+        extensions_to_patch = pin._config.get("_extensions_to_patch", None)
+        _config = pin._config
+        if extensions_to_patch:
+            _patch_extensions(extensions_to_patch)
+
+    c = traced_conn_cls(conn)
+
+    # if the connection has an info attr, we are using psycopg3
+    if hasattr(conn, "dsn"):
+        dsn = sql.parse_pg_dsn(conn.dsn)
+    else:
+        dsn = sql.parse_pg_dsn(conn.info.dsn)
+
+    tags = {
+        net.TARGET_HOST: dsn.get("host"),
+        net.TARGET_PORT: dsn.get("port", 5432),
+        db.NAME: dsn.get("dbname"),
+        db.USER: dsn.get("user"),
+        "db.application": dsn.get("application_name"),
+        db.SYSTEM: "postgresql",
+    }
+    Pin(tags=tags, _config=_config).onto(c)
+    return c
+
+
+def _patch_extensions(_extensions):
+    # we must patch extensions all the time (it's pretty harmless) so split
+    # from global patching of connections. must be idempotent.
+    for _, module, func, wrapper in _extensions:
+        if not hasattr(module, func) or isinstance(getattr(module, func), wrapt.ObjectProxy):
+            continue
+        wrapt.wrap_function_wrapper(module, func, wrapper)
+
+
+def _unpatch_extensions(_extensions):
+    # we must patch extensions all the time (it's pretty harmless) so split
```

Comment:
Nit: I like comments, but do we want the same one that we have above in `_patch_extensions`?